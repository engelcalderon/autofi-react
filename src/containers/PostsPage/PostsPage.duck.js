import { addEntities, addEntitiesRelationships } from '../../ducks/entities.duck';

export const POST_ENTITY_TYPE = 'post';
export const COMMENT_ENTITY_TYPE = 'comment';
export const COMMENT_POST_ENTITY_REL_NAME = 'comments';

export const FETCH_POSTS_REQUEST = 'app/PostsPage/FETCH_POSTS_REQUEST';
export const FETCH_POSTS_SUCCESS = 'app/PostsPage/FETCH_POSTS_SUCCESS';
export const FETCH_POSTS_ERROR = 'app/PostsPage/FETCH_POSTS_ERROR';

export const FETCH_COMMENTS_REQUEST = 'app/PostsPage/FETCH_COMMENTS_REQUEST';
export const FETCH_COMMENTS_SUCCESS = 'app/PostsPage/FETCH_COMMENTS_SUCCESS';
export const FETCH_COMMENTS_ERROR = 'app/PostsPage/FETCH_COMMENTS_ERROR';

const initialState = {
    postRefs: [],
    fetchPostsInProgress: false,
    fetchPostsError: null,
    fetchCommentsInProgress: false,
    fetchCommentsError: null
};

const postEntityRefs = data => {
    return data.map(d => ({ id: d.id, type: POST_ENTITY_TYPE }));
};

export default function postsPageReducer(state = initialState, action = {}) {
    const { type, payload } = action;
    switch (type) {
        case FETCH_POSTS_REQUEST:
            return {
                ...state,
                fetchPostsInProgress: true,
                fetchPostsError: null,
            };
        case FETCH_POSTS_SUCCESS:
            return {
                ...state,
                fetchPostsInProgress: false,
                postRefs: postEntityRefs(payload),
            };
        case FETCH_POSTS_ERROR:
            return {
                ...state,
                fetchPostsInProgress: false,
                fetchPostsError: payload.e
            };

        case FETCH_COMMENTS_REQUEST:
            return {
                ...state,
                fetchCommentsInProgress: true,
                fetchCommentsError: null,
            };
        case FETCH_COMMENTS_SUCCESS:
            return {
                ...state,
                fetchCommentsInProgress: false,
            };
        case FETCH_COMMENTS_ERROR:
            return {
                ...state,
                fetchCommentsInProgress: false,
                fetchCommentsError: payload.e
            };
        default:
            return state
    }
}

export const fetchPosts = () => (dispatch, getState, axios) => {
    dispatch({ type: FETCH_POSTS_REQUEST });
    return axios.get('posts')
        .then(response => {
            const data = response.data;
            dispatch(addEntities(POST_ENTITY_TYPE, data));
            dispatch({ type: FETCH_POSTS_SUCCESS, payload: data });
            return data;
        }).catch(err => dispatch({ type: FETCH_POSTS_ERROR, payload: { e: err } }))
};

export const fetchComments = () => (dispatch, getState, axios) => {
    dispatch({ type: FETCH_COMMENTS_REQUEST });
    return axios.get('comments')
        .then(response => {
            const data = response.data;
            dispatch(addEntities(COMMENT_ENTITY_TYPE, data));
            dispatch(addEntitiesRelationships(POST_ENTITY_TYPE, 'postId', COMMENT_POST_ENTITY_REL_NAME, data));
            dispatch({ type: FETCH_COMMENTS_SUCCESS });
            return data;
        }).catch(err => dispatch({ type: FETCH_COMMENTS_ERROR, payload: { e: err } }))
};

export const addComment = values => (dispatch, getState) => {
    const { entities } = getState().entities;
    const nextId = Object.keys(entities[COMMENT_ENTITY_TYPE]).length + 1;
    const comment = { id: nextId, ...values };
    return Promise.resolve()
        .then(() => {
            dispatch(addEntities(COMMENT_ENTITY_TYPE, [comment]));
            dispatch(addEntitiesRelationships(POST_ENTITY_TYPE, 'postId', COMMENT_POST_ENTITY_REL_NAME, [comment]));
        });
};