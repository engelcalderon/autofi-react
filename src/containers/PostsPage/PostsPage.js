import React from 'react';
import { array, object, bool } from 'prop-types';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { fetchPosts } from '../../ducks/posts.duck';
import { getEntities } from '../../ducks/entities.duck';
import { PostCard } from '../../components';

import css from './PostsPage.module.css';

export class PostsPageComponent extends React.Component {

    componentDidMount() {
        const { onFetchPosts } = this.props;
        onFetchPosts();
    }

    render() {
        const {
            posts,
            fetchPostsInProgress,
            fetchPostsError,
        } = this.props;

        if (fetchPostsInProgress) return <span>Loading...</span>

        if (fetchPostsError) return <span>Error fetching posts</span>

        return (
            <div className={css.page}>
                <div className={css.container}>
                    {posts.map(post => (
                        <PostCard
                            key={post.id}
                            post={post}
                        />
                    ))}
                </div>
            </div>
        )
    }
};


PostsPageComponent.defaultProps = {
    posts: [],
    fetchPostsError: null
}

PostsPageComponent.propTypes = {
    posts: array,
    fetchPostsInProgress: bool.isRequired,
    fetchPostsError: object,
}

const mapStateToProps = state => {
    const {
        postRefs,
        fetchPostsInProgress,
        fetchPostsError
    } = state.posts;

    const posts = getEntities(state, postRefs);

    return {
        posts,
        fetchPostsInProgress,
        fetchPostsError
    }
}

const mapDispatchToProps = dispatch => ({
    onFetchPosts: () => dispatch(fetchPosts()),
})

const PostsPage = compose(
    connect(
        mapStateToProps,
        mapDispatchToProps,
    ))(PostsPageComponent);

export default PostsPage;

