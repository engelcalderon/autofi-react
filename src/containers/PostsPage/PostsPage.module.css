body {
    background-color: lightblue;
}

.page {
    max-width: 1024px;
    margin: 0 auto;
}

.container {
    margin: 0 25px;
    padding: 24px 25px;
    display: flex;
    flex-direction: column;
    background-color: white;
    border-left: 1px solid gray;
    border-right: 1px solid gray; 
}