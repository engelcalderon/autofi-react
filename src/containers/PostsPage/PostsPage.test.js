import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { PostsPageComponent } from './PostsPage';

describe('PostsPage', () => {
    it('matches snapshot', () => {
        const posts = [
            {
                "postId": 1,
                "id": 1,
                "comments": [],
                "name": "id labore ex et quam laborum",
                "email": "Eliseo@gardner.biz",
                "body": "laudantium enim quasi est quidem magnam voluptate ipsam eos\ntempora quo necessitatibus\ndolor quam autem quasi\nreiciendis et nam sapiente accusantium"
            },
            {
                "postId": 1,
                "id": 2,
                "comments": [],
                "name": "quo vero reiciendis velit similique earum",
                "email": "Jayne_Kuhic@sydney.com",
                "body": "est natus enim nihil est dolore omnis voluptatem numquam\net omnis occaecati quod ullam at\nvoluptatem error expedita pariatur\nnihil sint nostrum voluptatem reiciendis et"
            },
            {
                "postId": 1,
                "id": 3,
                "comments": [],
                "name": "odio adipisci rerum aut animi",
                "email": "Nikita@garfield.biz",
                "body": "quia molestiae reprehenderit quasi aspernatur\naut expedita occaecati aliquam eveniet laudantium\nomnis quibusdam delectus saepe quia accusamus maiores nam est\ncum et ducimus et vero voluptates excepturi deleniti ratione"
            },
        ];
        const props = {
            posts,
            fetchPostsInProgress: false,
            fetchPostsError: {},
            onFetchPosts: () => { return Promise.resolve({}) },
            onFetchComments: () => () => { return Promise.resolve({}) },
            onAddComment: () => null
        }
        const tree = toJson(mount(<PostsPageComponent {...props} />), { mode: 'deep' });
        expect(tree).toMatchSnapshot();
    })
});