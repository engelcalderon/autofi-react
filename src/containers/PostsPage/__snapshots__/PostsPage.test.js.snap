// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostsPage matches snapshot 1`] = `
<span>
  Error fetching posts
</span>
`;
